crates/ibc/src/core/ics02_client/client_state.rs

```diff
@@ -42,14 +42,28 @@ pub trait ClientState:
     /// Latest height the client was updated to
     fn latest_height(&self) -> Height;
 
-    /// Freeze status of the client
-    fn is_frozen(&self) -> bool {
-        self.frozen_height().is_some()
+    /// Check if the given proof has a valid height for the client
+    fn validate_proof_height(&self, proof_height: Height) -> Result<(), ClientError> {
+        if self.latest_height() < proof_height {
+            return Err(ClientError::InvalidProofHeight {
+                latest_height: self.latest_height(),
+                proof_height,
+            });
+        }
+        Ok(())
     }
 
     /// Frozen height of the client
     fn frozen_height(&self) -> Option<Height>;
 
+    /// Assert that the client is not frozen
+    fn assert_not_frozen(&self) -> Result<(), ClientError> {
+        if let Some(frozen_height) = self.frozen_height() {
+            return Err(ClientError::ClientFrozen { frozen_height });
+        }
+        Ok(())
+    }
```

Comment:
As mentioned [here](https://github.com/cosmos/ibc-rs/pull/531#discussion_r1136151279), potentially there could be other error variants. thought, IMHO, I think returning a boolean is not a convention in places like here.

Reply:
> IMHO, I think returning a boolean is not a convention in places like here.

What makes you say that? Using `bool` definitely is idiomatic in Rust. For example, see [tokio](https://docs.rs/tokio/latest/tokio/io/struct.Ready.html#method.is_empty), [`std::vec::Vec::is_empty`](https://docs.rs/tokio/latest/tokio/io/struct.Ready.html#method.is_empty), [`Option::is_some`](https://doc.rust-lang.org/std/option/enum.Option.html#method.is_some), [`std::path::Path::is_file`](https://doc.rust-lang.org/std/path/struct.Path.html#method.is_file), etc etc etc

Reply:
Just to clarify, I didn't mean all over places in Rust. I was referring to _places like this_ and hoped we were on the same page from our [previous](https://github.com/cosmos/ibc-rs/issues/508#issuecomment-1460572951) discussion, so I didn't elaborate. Apologies if I confused you. 

Here are my two cents:

In the places `is_frozen` returns a `bool`, just afterward we only use that true/false to handle a failure. 
This resulted in:
1. Instead of only one implementation and simply using `?` to propagate the error, we have to take care the “true” condition (error variant) everywhere **(One vs Multiple impls)**
2. In action, having to handle multiple places, we easily can end up with inconsistent implementations (e.g. see [here](https://github.com/cosmos/ibc-rs/blob/45d5f0d25f323eb423d5511e2488be8104e1c39f/crates/ibc/src/core/ics02_client/handler/update_client.rs#L33-L35), and [here](https://github.com/cosmos/ibc-rs/blob/707182148a7df5916307af112128a6354dd9808b/crates/ibc/src/hosts/tendermint.rs#L30-L35)) and different yet similar error variants.(e.g. see [here](https://github.com/cosmos/ibc-rs/blob/088861bf573634d3179709e414b3176e7a893091/crates/ibc/src/core/ics04_channel/error.rs#L102-L103), [here](https://github.com/cosmos/ibc-rs/blob/088861bf573634d3179709e414b3176e7a893091/crates/ibc/src/core/ics04_channel/error.rs#L79-L80), [here](https://github.com/cosmos/ibc-rs/blob/d0d0683e9257fd83ef2e00897b29adf5e8445072/crates/ibc/src/clients/ics07_tendermint/error.rs#L88-L92), and [here](https://github.com/cosmos/ibc-rs/blob/088861bf573634d3179709e414b3176e7a893091/crates/ibc/src/core/ics02_client/error.rs#L24-L25)) **(Inconsistency)**

Further, since `assert_not_frozen` is only the by-product of misbehavior detection, I still see it as a chain and client-specific process. Builders should be able to include their own desirable error message to reflect the reason for the client freezing.

  3. This approach gives more control over their error handling. Otherwise, we should assume the reason/message or only can provide minimal info about the error. **(Customizable Errors)**
    

(Side note: I didn’t go through well-defined errors in this PR, since a serious and thorough evaluation is needed on that end) 

Also, note that the [examples](https://github.com/cosmos/ibc-rs/pull/531#discussion_r1136179670) you gave were implemented on **structs/enums** where end users can't have another implementation and just call and use them. Here, _in case of failures/errors_, if you don't want to impose your error system - unlike us :) - it's actually better to just return true/false and let users manage the error variant themselves.

One last thing, regarding your [rationale](https://github.com/cosmos/ibc-rs/pull/531#discussion_r1136179670). I agree with the height mandate, but still protocol does not restrict you to freeze a client only with height, and e.g. it is possible to set a state on a certain timestamp in the Solana ecosystem. 
That was why I liked the way `ibc-go` handle this and assume less about the clients giving a more generic client-neutral [`status()`](https://github.com/cosmos/ibc-go/blob/d840c699d988e32cb55f2c289d69f8ef2e5cea50/modules/core/exported/client.go#L52-L53) API. It was on my mind, so I asked you to investigate this in a later issue.

Reply:
I opened https://github.com/cosmos/ibc-rs/issues/536 for this. It even looks like we should also move away from `assert_not_frozen` method. If it's ok with you? here I only revise the naming and keep the rest to address in  #536.

Reply:
why not remove `frozen_height` as well?

Reply:
The changes we've had here were around the `validate_*` methods though inevitably we touched `is_fronzen` (bc of error variants). I am suggesting this since it is concerned with a different logical change that can also impact (remove) `confirm_not_frozen` and `expired` methods, for the sake of following the changes and a better changelog.

Reply:
We didn't *have* to touch `is_frozen`; I agree this would have been better handled in a different PR. I think it'd better to just revert the `assert_not_frozen()` changes; we expect to change it again, I am not convinced that this state is better than the previous, and this change is not covered in the changelog. But ultimately I think we had some good discussions around it, and we have a better idea of where we want to go with it.

Reply:
>  I am not convinced that this state is better than the previous

Note that changes around the `assert_not_frozen` differs. We ended up with the `assert_not_frozen` since changes in this PR required handling  **failures/errors** everywhere and I thought you agreed given the above reasoning. 

This change is different from the logical decision that should be made around keeping or dropping **status-related** interfaces (`asssert_not_frozen`, `frozen_height` and `is_expired`) or even gathering them under one interface named `status()` method.

> and different yet similar error variants.(e.g. see [here](https://github.com/cosmos/ibc-rs/blob/088861bf573634d3179709e414b3176e7a893091/crates/ibc/src/core/ics04_channel/error.rs#L102-L103), [here](https://github.com/cosmos/ibc-rs/blob/088861bf573634d3179709e414b3176e7a893091/crates/ibc/src/core/ics04_channel/error.rs#L79-L80), [here](https://github.com/cosmos/ibc-rs/blob/d0d0683e9257fd83ef2e00897b29adf5e8445072/crates/ibc/src/clients/ics07_tendermint/error.rs#L88-L92), and [here](https://github.com/cosmos/ibc-rs/blob/088861bf573634d3179709e414b3176e7a893091/crates/ibc/src/core/ics02_client/error.rs#L24-L25)) (Inconsistency)

Reverting `assert_not_frozen` would revert all the above error variants and everywhere we should use a couple of `map_err` and deal with this inconsistency